pkg/sources/file/file.go

```diff
@@ -0,0 +1,151 @@
+package file
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	diskbufferreader "github.com/bill-rich/disk-buffer-reader"
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/trufflesecurity/trufflehog/v3/pkg/common"
+	"github.com/trufflesecurity/trufflehog/v3/pkg/context"
+	"github.com/trufflesecurity/trufflehog/v3/pkg/handlers"
+	"github.com/trufflesecurity/trufflehog/v3/pkg/pb/source_metadatapb"
+	"github.com/trufflesecurity/trufflehog/v3/pkg/pb/sourcespb"
+	"github.com/trufflesecurity/trufflehog/v3/pkg/sanitizer"
+	"github.com/trufflesecurity/trufflehog/v3/pkg/sources"
+)
+
+// Source is a file entity.
+type Source struct {
+	sourceId, jobId, fileSize int64
+	verify                    bool
+	name, path, fileName      string
+	aCtx                      context.Context
+	log                       *log.Entry
+	sources.Progress
+}
+
+// Ensure the Source satisfies the interface at compile time.
+var _ sources.Source = (*Source)(nil)
+
+// Type returns the type of the source.
+func (s *Source) Type() sourcespb.SourceType {
+	return sourcespb.SourceType_SOURCE_TYPE_FILE
+}
+
+// SourceID returns the source ID.
+func (s *Source) SourceID() int64 {
+	return s.sourceId
+}
+
+// JobID returns the job ID.
+func (s *Source) JobID() int64 {
+	return s.jobId
+}
+
+// Init returns an initialized File source.
+func (s *Source) Init(aCtx context.Context, name string, jobId, sourceId int64, verify bool, connection *anypb.Any, _ int) error {
+	s.log = log.WithField("source", s.Type()).WithField("name", name)
+
+	s.aCtx = aCtx
+	s.name = name
+	s.sourceId = sourceId
+	s.jobId = jobId
+	s.verify = verify
+
+	var conn sourcespb.File
+	if err := anypb.UnmarshalTo(connection, &conn, proto.UnmarshalOptions{}); err != nil {
+		return fmt.Errorf("failed to unmarshal connection: %w", err)
+	}
+
+	s.path = conn.Path
+	// Set the path to stdin if one is not provided.
+	if s.path == "" {
+		s.path = os.Stdin.Name()
+	}
+	s.fileName = filepath.Base(s.path)
+
+	fi, err := os.Stat(s.path)
+	if err != nil {
+		return fmt.Errorf("failed to stat file: %w", err)
+	}
+	if !fi.Mode().IsRegular() {
+		return fmt.Errorf("file is not a regular file: %v", s.path)
+	}
```

Comment:
**Suggestion:** Maybe check if the file is a directory here instead? Otherwise I think this check prevents scanning symlinks, which I could see a legitimate use for.